ipq40xx: 5.4: backport MDIO, add QCA807x PHY driver and update EDMA

robimarko opened the pull request:
This patch series backports upstreamed MDIO driver, adds QCA807x PHY driver as well as some EDMA improvements.

A proper QCA807x PHY driver is introduced along with its DT bindings and DTS changes needed.
Fixups needed for upstream MDIO driver as well as new PHY driver are included.

PHY driver includes all of the features that both Linux and PHY itself support except for WoL(As that makes no sense),
it also includes cable testing but that will have to wait until kernel 5.8 since netlink support for ethool landed in 5.7 and cable testing framework in 5.8.

Some EDMA improvements now that a proper QCA807x PHY driver is available are also added.

PHY driver will be sent upstream soon.

robimarko committed:
ipq40xx: edma: convert to of_mdio_find_bus()

With the reworked MDIO driver, EDMA will fail to get the MII BUS as it
used the MII BUS stored inside the MDIO structure private data.

This obviously does not work with the modernized driver, so lets switch
to using a purpose build of_mdio_find_bus() which will return the MII
BUS and only requires the MDIO node to be passed.
This is easy as we already have the node parsed.

Also, since we now require OF_MDIO add that as dependency.

Signed-off-by: Robert Marko <robert.marko@sartura.hr>

robimarko committed:
ipq40xx: net: ethernet: edma: fix link detection

PHY needs to be soft reset before starting it from ethernet driver as
AR40xx calibration will leave it in unwanted state.

Signed-off-by: Robert Marko <robert.marko@sartura.hr>

robimarko committed:
ipq40xx: net: phy: ar40xx: remove PHY handling

Since we now have proper PHY driver for the QCA807x PHY-s, lets remove
PHY handling from AR40xx.

This removes PHY driver, PHY GPIO driver and PHY init code.
AR40xx still needs to handle PSGMII calibration as that requires R/W
from the switch, so I am unable to move it into PHY driver.

This also converted the AR40xx driver to use OF_MDIO to find the MDIO
bus as it now cant be set through the PHY driver.
So lets depend on OF_MDIO in KConfig.

Signed-off-by: Robert Marko <robert.marko@sartura.hr>

robimarko committed:
ipq40xx: dts: add QCA807x properties

This adds necessary DT properties for QCA807x PHY-s to IPQ4019 DTSI.

Also adds the PSGMII PHY as it wont get probed otherwise.

Signed-off-by: Robert Marko <robert.marko@sartura.hr>

robimarko commented:
@adschm Something for you to take a look at.

I would also be glad if people can test this on their boards as well, especially on boards that use RGMII and not QCA807x PHY-s.
I dont have any of those.

adschm commented on target/linux/ipq40xx/config-5.4:
```diff
@@ -414,6 +414,7 @@ CONFIG_PINCTRL_MSM=y
 # CONFIG_PINCTRL_SDM660 is not set
 # CONFIG_PINCTRL_SDM845 is not set
 # CONFIG_PINCTRL_SM8150 is not set
+CONFIG_PLUGIN_HOSTCC="g++"
```
I think this line should be dropped.

robimarko replied:
Yeah, dropped it finally.

adschm commented:
I think it's a little confusing that you remove the ar40xx driver from patches in commit 1 and then add it to files in commit 4. I actually wondered why it was dropped without a comment when reviewing commit 1 ...

Personally, I would just have the move from "patches" to files in a dedicated commit, which might be best put before the other commits here. And I'd actually remove the (relevant part of the) patch from 4.19 _and_ 5.4, and add kernel version switches for the few lines, so the driver can be put into files, not files-5.4.

robimarko commented:
I agree that I made it confusing as I removed it when updating the patches.
I will clear that up a bit.

Can we simply drop 4.19?
It serves no purpose, and I don't plan on adding kernel switches for linkmode conversion and later PHY

adschm commented:
> Can we simply drop 4.19?
> It serves no purpose, and I don't plan on adding kernel switches for linkmode conversion and later PHY conversion into AR40xx.

I actually thought about that as well. Essentially, the old kernels are kept for regression testing. I plan to propose their removal during 20.xx rc stage.

But you are probably right that it's not worth to complicate anything because of that. So, I will just pick the move from patches-5.4 to files-5.4 right away, as this should be easiest at the moment.

adschm commented on target/linux/ipq40xx/files-5.4/drivers/net/phy/ar40xx.h:
```diff
@@ -0,0 +1,342 @@
+ * OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+ #ifndef __AR40XX_H
```
This is actually indented (like in the patch). Maybe one should fix this in a separate patch as well.

robimarko replied:
Hm, I applied the patches and then copied ar40xx.c and ar40xx.h from the patched kernel tree, so its gotta be exactly the same as in the patch

adschm replied:
The space is also there in the patch, it's not your fault. Nevertheless, this looks broken to me (as #define only works at the beginning of the line?), so I wanted to ask for your opinion ...

robimarko replied:
Yeah, space should not be here.
It most likely works as the preprocessor will take care of that anyway

adschm commented:
I just merged the ar40xx move separately.

adschm commented on target/linux/ipq40xx/patches-5.4/902-dts-ipq4019-ap-dk04.1.patch:
```diff
@@ -116,17 +117,17 @@
  
--		pci@40000000 {
--			status = "ok";
--			perst-gpio = <&tlmm 38 0x1>;
```
This looks like a _change_ to me, i.e. not removing pci@40000000 anymore?

robimarko replied:
Yes, it made no sense in dropping the PCI node as DK04.1 has PCI-E slot as default.

adschm replied:
I essentially just removed it when migrating from files to patch in order to have the exact same nodes (so, I didn't need to test it). While I generally agree that it's probably more correct to have the pci node, I wouldn't add it as a by-product here, but would prefer to have it in a separate commit.

robimarko replied:
Ok, will move it as a separate thing

DavideFioravanti commented:
> I would also be glad if people can test this on their boards as well, especially on boards that use RGMII and not QCA807x PHY-s.

I am testing this on a [Netgear SRS60](https://github.com/openwrt/openwrt/pull/3469) (unfortunately I don't have any ipq40xx using RGMII). Seems to work correctly (no kernel logs, no connection drops) and solves the problem mentioned [here](https://github.com/openwrt/openwrt/pull/3491). I will keep testing it.

robimarko commented:
> I just merged the ar40xx move separately.

Great, will rebase this tomorrow.

robimarko commented:
> > I would also be glad if people can test this on their boards as well, especially on boards that use RGMII and not QCA807x PHY-s.
> 
> I am testing this on a [Netgear SRS60](https://github.com/openwrt/openwrt/pull/3469) (unfortunately I don't have any ipq40xx using RGMII). Seems to work correctly (no errors in the logs, no connection drops) and solves the problem mentioned [here](https://github.com/openwrt/openwrt/pull/3491). I will keep testing it.

Great, thanks for testing.
Nice to see that your issue is finally solved.

robimarko commented:
Rebased this to resolve the conflict.

dengqf6 commented on target/linux/ipq40xx/files-5.4/drivers/net/phy/qca807x.c:
```diff
@@ -0,0 +1,619 @@
+	return 0;
+}
+
+static int qca807x_gpio(struct phy_device *phydev)
```
How about register it as an LED controller instead of GPIO?

robimarko replied:
Well, I thought about it at first.
Unfortunately, I have seen vendors like Mikrotik use these like regular GPIO-s to enable stuff like the power to USB or SFP.
So it seems more proper to have it be like a generic GPIO, plus that allows simply using gpio-leds in DT.

dengqf6 replied:
I see. It would be better if we could utilize its hardware blink feature though.

robimarko replied:
I agree, that is why this is fully optional and only used for cases where vendor did not correctly wire the LED-s or even worse used these for all kinds of stuff

dengqf6 replied:
Maybe register it selectively? LED controller by default, GPIO controller if needed?
Could also be used to handle single LED.

robimarko replied:
I am trying to avoid having Linux do anything at all if not needed as if we register them as LED controllers Linux will take over.
That would break existing behaviour and we won't have triggers for LED-s that are not attached to EDMA directly.

dengqf6 commented on target/linux/ipq40xx/files-5.4/drivers/net/phy/qca807x.c:
```diff
@@ -0,0 +1,619 @@
+
+	gc->label = dev_name(dev);
+	gc->base = -1;
+	gc->ngpio = 4;
```
Each port only has 2 LEDs, so it should be 2

robimarko replied:
Yeah, just checked the datasheet.
You are correct, there are only 2 status LED-s per PHY.
The rest of the digital pins are MDC/MDIO, reset, interrupts, etc.

I was confused by the existing driver in AR40xx.
This is gonna make converting the bindings on Fritzbox 4040 impossible without having the device.

dengqf6 commented on target/linux/ipq40xx/files-5.4/drivers/net/phy/qca807x.c:
```diff
@@ -314,7 +322,7 @@ static int qca807x_gpio_get_direction(struct gpio_chip *gc, unsigned int offset)
 static int qca807x_gpio_get_reg(unsigned int offset)
 {
-	return QCA807X_MMD7_LED_100N + offset % 4;
+	return QCA807X_MMD7_LED_100N_1 + offset % 4;
```
Only `QCA807X_MMD7_LED_100N_2` and `QCA807X_MMD7_LED_1000N_2` are used to force LED output, so it should be like
```c
return QCA807X_MMD7_LED_100N_2 + (offset % 2) * 2;
```

robimarko replied:
Will correct those.

robimarko commented on target/linux/ipq40xx/files-5.4/drivers/net/phy/qca807x.c:
```diff
@@ -485,6 +493,42 @@ static int qca807x_ack_intr(struct phy_device *phydev)
+	bool led_config = false;
+	int val;
+
+	val = phy_read_mmd(phydev, MDIO_MMD_AN, QCA807X_MMD7_LED_1000N_1);
```
Well, I have not encountered that yet, but your remark is correct as it should be handled as well.
I can probably think of some DT bindings to cover that as well.

robimarko replied:
This is left to be done, I can think of a simple dt bindings that could achieve this.
I want to avoid too much new bindings.

dengqf6 commented on target/linux/ipq40xx/files-5.4/arch/arm/boot/dts/qcom-ipq4018-fritzbox-4040.dts:
```diff
@@ -110,33 +100,33 @@
 		wlan {
 			label = "green:wlan";
-			gpios = <&qca8075 1 GPIO_ACTIVE_HIGH>;
+			gpios = <&ethphy0 1 GPIO_ACTIVE_HIGH>;
```
As you mentioned, after correcting `gc->ngpio` , `1 (0x8074 + 1)` here should be `0 (0x8075 + 0 * 2)`

robimarko replied:
Thanks, this should work as on AP120C the conversion works.

dengqf6 commented on target/linux/ipq40xx/files-5.4/arch/arm/boot/dts/qcom-ipq4018-fritzbox-4040.dts:
```diff
@@ -110,33 +100,33 @@
 		panic: info_red {
 			label = "red:info";
-			gpios = <&qca8075 3 GPIO_ACTIVE_HIGH>;
+			gpios = <&ethphy0 3 GPIO_ACTIVE_HIGH>;
```
And `3 (0x8074 + 3)` here should be `1 (0x8075 + 1 * 2)`

robimarko commented:
Pushed changes to address @dengqf6 remarks(Well, almost all of them) and patch conflicts.
@adschm Can we at least drop 4.19 for IPQ40xx if not for all targets?

adschm commented:
@robimarko This guy here specifically reported problems with ipq40xx and 5.4:

https://www.mail-archive.com/openwrt-devel@lists.openwrt.org/msg55205.html

For the moment, I think we should keep 4.19 until the two problems named there are resolved.

robimarko commented:
UBNT devices he mentioned are ath79 and the "switch" issue is due to this:
https://github.com/openwrt/openwrt/commit/9da2b567605b0964d921b9ca4f0c9886db4f636d#diff-043b5b3b3004fa40ab2e96f33c6be34d6dd190b4b281ad40e320a2cf40504b74

It has been confirmed in multiple places on the forum.

adschm commented:
> [...] and the "switch" issue is due to this:
> [9da2b56#diff-043b5b3b3004fa40ab2e96f33c6be34d6dd190b4b281ad40e320a2cf40504b74](https://github.com/openwrt/openwrt/commit/9da2b567605b0964d921b9ca4f0c9886db4f636d#diff-043b5b3b3004fa40ab2e96f33c6be34d6dd190b4b281ad40e320a2cf40504b74)
> 
> It has been confirmed in multiple places on the forum.

So, 4.19 is "better" just because the relevant patch has only been applied to 5.4?

robimarko commented:
Yeah, looks like it, small reference to the forum thread:
https://forum.openwrt.org/t/ipq40xx-low-performance-when-sending-traffic/74520/6

robimarko commented:
Great, will move stuff back to files folder

robimarko committed:
ipq40xx: dts: convert PHY GPIO bindings

Since the new PHY driver manages each PHY individually and therefore 
registers each PHY that is marked with gpio-controller; DT property as a
GPIO controller we need to convert old DT bindings to account for this.

Only 2 boards use this so its not much of an issue.

Signed-off-by: Robert Marko <robert.marko@sartura.hr>

robimarko committed:
ipq40xx: add Qualcomm QCA807x driver

This adds driver for the Qualcomm QCA8072 and QCA8075 PHY-s.

They are 2 or 5 port IEEE 802.3 clause 22 compliant
10BASE-Te, 100BASE-TX and 1000BASE-T PHY-s.

They feature 2 SerDes, one for PSGMII or QSGMII connection with MAC,
while second one is SGMII for connection to MAC or fiber.

Both models have a combo port that supports 1000BASE-X and 100BASE-FX
fiber.

Each PHY inside of QCA807x series has 2 digitally controlled output only
pins that natively drive LED-s.
But some vendors used these to driver generic LED-s controlled by
user space, so lets enable registering each PHY as GPIO controller and
add driver for it.

This also adds the ability to specify DT properties so that 1000 Base-T
LED will also be lit up for 100 and 10 Base connections.

This is usually done by U-boot, but boards running mainline U-boot are
not configuring this yet.

These PHY-s are commonly used in Qualcomm IPQ40xx, IPQ60xx and IPQ807x
boards.

Signed-off-by: Robert Marko <robert.marko@sartura.hr>

robimarko commented:
Pushed the updated version to deal with 4.19 drop.

robimarko commented:
@adschm Any chance of merging this?

I should have fiber working in the next couple of weeks as well.

robimarko commented:
Rebased and refreshed the patches again.

@adschm Is there anything else preventing this from being merged?
There are 2 PR-s already depending on this.

robimarko commented:
I rebased this once again.
@adschm Anything preventing this from being merged?

adschm commented:
This goes a little too deep for my ipq40xx competence; I'd prefer to have this dealt with by another committer unless I have a big amount of time at some point or you can get a Reviewed-by and several Tested-by ...

robimarko commented:
I can probably get some Tested-by-s several people already tested this.
Reviewed-by on the other hand would most likely require a committer to review this, and I personally don't see any interest in IPQ40xx which is kind of weird as its one of the most active platforms.

f00b4r0 commented:
Tested-by me on hap-ac2 (ipq40xx). Not reviewed though :)

DavideFioravanti commented:
I can confirm it's working and solves the WAN problem for this PR https://github.com/openwrt/openwrt/pull/3469

I don't have the knowledge to review this, but I can surely add my `tested-by`:
`Tested-by: Davide Fioravanti <pantanastyle@gmail.com>`

adschm commented:
Reviewed-by is essentially acceptable from anyone, although I of course will have a look on who gave it and whether that person can be expected to actually have done a proper review.

Actually, the entire process would be much faster if non-committers would dare to provide Reviewed-bys more often, as this would make it much easier for committers to merge stuff they are only 70 % sure about.
There are many specialists around here that are not committers. Those typically have valuable insights into specific areas (like yourself), while committers might have more overview. IMO having non-committer reviews more often would combine these two angles towards reaching more competent and quick decisions.

subixonfire commented:
You can add: Tested-by:Alen Opačič <subixonfire@gmail.com>

adschm commented:
> You can add: Tested-by:Alen Opačič [subixonfire@gmail.com](mailto:subixonfire@gmail.com)

Please state the device you tested on.

f00b4r0 commented:
@adschm makes sense indeed.
@robimarko this is a fairly large patchset, it would certainly help to know which of these changes are pure backports, which have been sent upstream (and possibly reviewed there) and which are "for us only": the review effort for each category is quite different IMHO (next to nothing for the former, a lot more for the latter). This would also help potential reviewer to know where to focus their attention.

I can generally review kernel code up to a point, the caveat here being that I'm not familiar with this particular platform.

HTH

subixonfire commented:
> > You can add: Tested-by:Alen Opačič [subixonfire@gmail.com](mailto:subixonfire@gmail.com)
> 
> Please state the device you tested on.

Sorry, tested on Mikrotik hAP ac2

adschm commented:
> and I personally don't see any interest in IPQ40xx which is kind of weird as its one of the most active platforms.

I cannot comment on what others do.

Personally, although I'm aware this doesn't help you with your problem, I'm not really that familiar with the ipq40xx target; I mostly got in touch when doing reorganization/cleanup of "generic" stuff. So, switching between drivers on this platform is not really my call, but should be evaluated by somebody familiar with the particular environment.

adschm commented:
BTW, following what @f00b4r0 stated, an additional complexity is created by mixing the kernel patch "updates" in the first commit with the introduction of the IPQ4019-MDIO driver. While I'm fine with the other rearrangements there (and actually reviewed them), I cannot do the same for the mdio driver.

@robimarko Would it be an option for you to split the mdio driver introduction from the other changes ("take upstream versions"), so I can merge the latter and we have them out of the way for reviewers? Or are these changes interdependent?

I'd be fine with a link to a branch in your repo or similar with only the "take upstream" changes as a single patch, don't need to update the entire PR for that. This might be a small step making the big step easier to achieve.

robimarko commented:
@f00b4r0 I agree that it is a large patchset, but it's long overdue.
As far as backported stuff goes, it's easy to spot the patches as they all start with 0 and have a kernel version from which they were backported also.
The rest of the changes to the patches was reordering them as we had a lot of patches that are not backports at 0 and others at random numbering.

PHY driver I introduce here is not yet sent upstream as I was working on SFP support for it as well, that is close to being done and then I will send the driver upstream along with DT bindings for it.
That was my plan from the beginning and I would update the PHY driver in OpenWrt with the upstreamed version after that is done.

Changes exclusive to OpenWrt are EDMA and AR40xx ones, these don't exist upstream.

@adschm Sure, I can split out the MDIO patch, but it just replaces our downstream version with the upstreamed one.
I thought about splitting the patch changes further, but the issue is that they all depend on each other in a specific ordering and its a pain due to the current mess.

EnfermeraSexy commented:
I have 3 units of the Mikrotik hAP ac2 (ipq40xx) and every unit is working without problems and stable.

adschm commented:
> @adschm Sure, I can split out the MDIO patch, but it just replaces our downstream version with the upstreamed one.
> I thought about splitting the patch changes further, but the issue is that they all depend on each other in a specific ordering and its a pain due to the current mess.

The problem is that the MDIO patch also triggers the adjustments in the subsequent patches etc. (e.g. "ipq40xx: edma: convert to of_mdio_find_bus()" that requires specific knowledge). So, it would be nice to just have anything off the table where we can simply take the upstream version without any adjustments necessary outside patches directory.

f00b4r0 commented:
@robimarko I know it can be frustrating, but it's kind of a chicken-and-egg problem: large patchset means it's harder to review, takes more time, gets old and "overdue". That's why kernel maintainers require changes to be broken down into small coherent chunks. It makes reviewing/assessing the code easier.

I would also not wait to send upstreamable changes upstream: if code gets vetted there, that's less work for reviewers here. And if it gets a NAK, it's also good to know early so it can be adjusted: if OpenWRT merges code that's NAKd upstream, it's more work for everyone. (NB: remember to adjust your commit message line wrap :)

Still, thanks again for your efforts.

adschm commented:
And BTW I personally won't cry if the patch numbers are not consecutive.

robimarko committed:
ipq40xx: backport upstream MDIO driver

IPQ40xx MDIO driver was upstreamed in kernel version 5.8.
So lets backport the upstream version and drop our local one.

This also refreshed the kernel config since the symbol name has changed.

Signed-off-by: Robert Marko <robert.marko@sartura.hr>

robimarko commented:
@adschm I have separated out the MDIO driver backport.
So now other backports and reordering can be picked while not requiring changes to anything else.

@f00b4r0 I understand, but the PHY driver requires a number of changes for it to work and I tried to separate those into separate commits.

I can't really send it upstream before I finish it, like I said it's 99% done and will be sent next week.
I would urge you to take a look at the state of the current PHY "driver" in AR40xx.

adschm commented on target/linux/ipq40xx/patches-5.4/0003-v5.7-ARM-dts-qcom-add-gpio-ranges-property.patch:
```diff
@@ -0,0 +1,119 @@
+From 8b99dc0922618062a1589ebd74df6108b4f9ac22 Mon Sep 17 00:00:00 2001
+From: Christian Lamparter <chunkeey@gmail.com>
+Date: Wed, 8 Jan 2020 13:54:55 +0100
+Subject: [PATCH] ARM: dts: qcom: add gpio-ranges property
```
Off-topic here, but I wonder whether that one should be moved to generic patches at some point?

robimarko replied:
Since it affects only 2 targets that are in OpenWrt I don't know if it makes sense.
It's gonna be dropped at the next kernel update anyway.

adschm commented on target/linux/ipq40xx/patches-5.4/0017-v5.8-phy-add-driver-for-Qualcomm-IPQ40xx-USB-PHY.patch:
```diff
@@ -137,15 +124,6 @@ Signed-off-by: John Crispin <john@phrozen.org>
-+	writel(PHY_CTRL3_VAL, phy->base + PHY_CTRL3_ADDR);
-+	writel(PHY_CTRL4_VAL, phy->base + PHY_CTRL4_ADDR);
-+	writel(PHY_MISC_VAL, phy->base + PHY_MISC_ADDR);
-+	writel(PHY_IPG_VAL, phy->base + PHY_IPG_ADDR);
```
What happened to the magic values?

robimarko replied:
They are not needed at all, they were only for preproduction HW.

Sartura has been upstreaming all of this throughout the last and this year as we use mainline kernels.
So, everything is being used actively.

adschm commented:
Thanks, I've pushed the first patch.

robimarko commented:
> Thanks, I've pushed the first patch.

Thanks, hopefully, we can get the rest over the line as well.

john-tho commented:
This PR wants a rebase for the 5.4.77 kernel bump https://github.com/openwrt/openwrt/commit/667f6c7f49c94213ca43a42ad5a9e23abfd81861

@chunkeey @blogic It looks like you have had a bit to do with IPQ40xx edma, any comments on this PR?

To clear this up for me:
This PR introduces an upstreamed MDIO driver, and also a PHY driver for a common IPQ40xx transceiver that is planned for upstream?
Otherwise, Openwrt is using [stale drivers for IPQ40xx SoC components that are not supported upstream](https://github.com/openwrt/openwrt/pull/1904#issuecomment-471003615):
- [switch ar40xx](https://groups.google.com/a/chromium.org/g/chromium-os-reviews/c/L-QqqpBjNak)
- [ethernet ess e dma?](https://groups.google.com/a/chromium.org/g/chromium-os-reviews/c/fD2ezuw_vWI)

f00b4r0 commented on target/linux/ipq40xx/files/drivers/net/phy/qca807x.c:
```diff
@@ -0,0 +1,672 @@
+
+		/* Prevent PSGMII going into hibernation via PSGMII self test */
+		psgmii_serdes = phy_read_mmd(phydev, MDIO_MMD_PCS, PSGMII_MMD3_SERDES_CONTROL);
+		psgmii_serdes &= (~(1<<1));
```
~BIT(1)? Maybe even #define the bit or the mask clearly?

robimarko replied:
Yeah, missed this one.
I can't really define it as this register is fully undocumented.

f00b4r0 commented on target/linux/ipq40xx/files/drivers/net/phy/qca807x.c:
```diff
@@ -0,0 +1,672 @@
+	}
+
+	/* Optionally configure LED-s */
+	#ifdef CONFIG_GPIOLIB
```
Here and in other places, use `if (IS_ENABLED(CONFIG_GPIOLIB))` instead.
See https://www.kernel.org/doc/html/latest/process/coding-style.html#conditional-compilation

robimarko replied:
Thanks, will do.

robimarko commented:
Rebased and pushed to resolve the conflicts.

This PR has replaced all of the local patches with the upstreamed versions if accepted, that includes backporting the upstream MDIO driver, driver for the common QCA807x PHY-s as well as applying fixes to AR40xx and ESSEDMA to make them work with the new drivers.

f00b4r0 reviewed (commented):
Quick read through qca807x.c, sorry to be nitpicking but the file seems to have some formatting issues (indenting and odd line breaks) for the first half. Beware of conditional compiling best practices, and a few missing spaces here and there between `if` and opening `(`. Otherwise fairly easy to read. Can't comment on the actual code as I don't know the hardware. HTH

adschm commented on target/linux/ipq40xx/files/drivers/net/phy/qca807x.c:
```diff
@@ -0,0 +1,672 @@
+#ifdef CONFIG_GPIOLIB
+static int qca807x_gpio_get_direction(struct gpio_chip *gc, unsigned int offset)
+{
+	#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,5,0)
+	return GPIO_LINE_DIRECTION_OUT;
+	#else
+	return GPIOF_DIR_OUT;
+	#endif
```
> Since this is a new driver, are these necessary?

Assuming this is used on 5.4 as well (as intended with the current patch state), the switch will be needed as soon as we introduce 5.10. (for OpenWrt, of course not for inclusion upstream)

f00b4r0 replied:
> (for OpenWrt, of course not for inclusion upstream)

My point exactly. I understand we (OpenWRT) may need those, but for upstreaming I'm worried this will get frowned upon.

robimarko replied:
All of the version switches will not be in the version submitted for upstreaming.
I only added these for OpenWrt because of kernel 5.4, I am working on all of this under 5.10 at work.
I actually got SFP working properly today, so after I test that it works properly on the new set of boards I have coming next week I will push it upstream.
In the meantime, I will clean it up and write docs.

f00b4r0 replied:
IMHO #ifdefs should not be indented

lynxis replied:
+1

robimarko committed:
ipq40xx: net: ethernet: edma: use generic ksettings functions

Since we now have a proper PHY driver for QCA807x and AR803x has already
been supported properly there is no need for the driver to be poking
on PHY registers for ethtool ops.

So, lets simply use the generic
phy_ethtool_ksettings_get/phy_ethtool_ksettings_set functions.

This also has the advantage of properly populating stuff other than
speeds like, transceiver type, MDI-X etc.

ethtool before:
root@OpenWrt:/# ethtool eth1
Settings for eth1:
        Supported ports: [ TP MII ]
        Supported link modes:   10baseT/Half 10baseT/Full
                                100baseT/Half 100baseT/Full
                                1000baseT/Full
                                1000baseX/Full
        Supported pause frame use: Symmetric Receive-only
        Supports auto-negotiation: Yes
        Supported FEC modes: Not reported
        Advertised link modes:  10baseT/Half 10baseT/Full
                                100baseT/Half 100baseT/Full
                                1000baseT/Full
                                1000baseX/Full
        Advertised pause frame use: Symmetric Receive-only
        Advertised auto-negotiation: Yes
        Advertised FEC modes: Not reported
        Link partner advertised link modes:  10baseT/Half 10baseT/Full
                                             100baseT/Half 100baseT/Full
                                             1000baseT/Full
        Link partner advertised pause frame use: No
        Link partner advertised auto-negotiation: No
        Link partner advertised FEC modes: Not reported
        Speed: 1000Mb/s
        Duplex: Full
        Port: Twisted Pair
        PHYAD: 4
        Transceiver: internal
        Auto-negotiation: on
        MDI-X: Unknown
        Supports Wake-on: d
        Wake-on: d
        Current message level: 0x00000000 (0)

        Link detected: yes

ethtool after:
root@OpenWrt:/# ethtool eth1
Settings for eth1:
        Supported ports: [ TP MII ]
        Supported link modes:   10baseT/Half 10baseT/Full
                                100baseT/Half 100baseT/Full
                                1000baseT/Full
                                1000baseX/Full
        Supported pause frame use: Symmetric Receive-only
        Supports auto-negotiation: Yes
        Supported FEC modes: Not reported
        Advertised link modes:  10baseT/Half 10baseT/Full
                                100baseT/Half 100baseT/Full
                                1000baseT/Full
                                1000baseX/Full
        Advertised pause frame use: Symmetric Receive-only
        Advertised auto-negotiation: Yes
        Advertised FEC modes: Not reported
        Link partner advertised link modes:  10baseT/Half 10baseT/Full
                                             100baseT/Half 100baseT/Full
                                             1000baseT/Full
        Link partner advertised pause frame use: Symmetric Receive-only
        Link partner advertised auto-negotiation: Yes
        Link partner advertised FEC modes: Not reported
        Speed: 1000Mb/s
        Duplex: Full
        Port: Twisted Pair
        PHYAD: 4
        Transceiver: external
        Auto-negotiation: on
        MDI-X: off (auto)
        Supports Wake-on: d
        Wake-on: d
        Current message level: 0x00000000 (0)

        Link detected: yes

Signed-off-by: Robert Marko <robert.marko@sartura.hr>

robimarko committed:
ipq40xx: net: ethernet: edma: use generic PHY print

Lets use the generic upstream phy_print_status() instead of doing 
something similar by hand.

Before:
ess_edma c080000.edma: eth1: GMAC Link is up with phy_speed=1000

After:
ess_edma c080000.edma eth1: Link is Up - 1Gbps/Full - flow control rx/tx

Signed-off-by: Robert Marko <robert.marko@sartura.hr>

robimarko commented:
@f00b4r0 @adschm I have updated rebased the patch series, added some more EDMA improvements as well as added fiber/SFP support to the PHY driver along with some cosmetic changes.

This PHY driver version will be the one sent upstream.

f00b4r0 commented on target/linux/ipq40xx/files/drivers/net/phy/qca807x.c:
```diff
@@ -0,0 +1,828 @@
+
+	if (!cnt) {
+		ret = phy_clear_bits(phydev, MII_NWAYTEST,
+							 QCA807X_SMARTSPEED_EN);
```
Indentation of line break

f00b4r0 commented on target/linux/ipq40xx/files/drivers/net/phy/qca807x.c:
```diff
@@ -0,0 +1,828 @@
+}
+
+static int qca807x_get_tunable(struct phy_device *phydev,
+							   struct ethtool_tunable *tuna, void *data)
```
Indentation

f00b4r0 commented on target/linux/ipq40xx/files/drivers/net/phy/qca807x.c:
```diff
@@ -0,0 +1,828 @@
+}
+
+static int qca807x_set_tunable(struct phy_device *phydev,
+							   struct ethtool_tunable *tuna, const void *data)
```
Same

f00b4r0 commented on target/linux/ipq40xx/files/drivers/net/phy/qca807x.c:
```diff
@@ -0,0 +1,828 @@
+	{
+		PHY_ID_MATCH_EXACT(PHY_ID_QCA8072),
+		.name           = "Qualcomm QCA8072",
+		#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,8,0)
+		.flags			= PHY_POLL_CABLE_TEST,
+		#endif
```
I wouldn't indent #if 
same below.

lynxis replied:
+1

f00b4r0 commented on target/linux/ipq40xx/files/drivers/net/phy/qca807x.c:
```diff
@@ -0,0 +1,828 @@
+		#endif
+		/* PHY_GBIT_FEATURES */
+		.probe 			= qca807x_probe,
+		.config_init 	= qca807x_config,
+		.read_status	= qca807x_read_status,
+		.config_intr	= qca807x_config_intr,
+		.ack_interrupt	= qca807x_ack_intr,
+		.soft_reset		= genphy_soft_reset,
+		.get_tunable	= qca807x_get_tunable,
```
mixed indentation.
Same below

lynxis replied:
+1

f00b4r0 reviewed (commented):
Purely styling review, you still have a lot of indentation issues.

I'd suggest either don't wrap some code lines which aren't very long or indent correctly. 
See https://www.kernel.org/doc/html/latest/process/coding-style.html#breaking-long-lines-and-strings
Bear in mind that kernel style is 8-column _tabs_.
When aligning struct members or #define values, I'd also recommend being consistent.
Generally speaking, I'd recommend running checkpatch.pl and making sure you can justify all remaining violations.

Also bear in mind that commit message body must still be wrapped at 75 columns.

Proper styling will ease upstream review and prevent a preemptive NAK (were your patch won't even be read), in my experience. HTH

lynxis commented on target/linux/ipq40xx/config-5.4:
```diff
@@ -4,22 +4,6 @@ CONFIG_ALIGNMENT_TRAP=y
 CONFIG_AR40XX_PHY=y
 CONFIG_ARCH_32BIT_OFF_T=y
 CONFIG_ARCH_CLOCKSOURCE_DATA=y
-CONFIG_ARCH_HAS_BINFMT_FLAT=y
```
Please drop all symbol modifications except the 4019 symbol you need.

robimarko replied:
Done, I separated the kernel refresh and MDIO symbols.

lynxis commented:
I oiverlooked this PR while reviewing https://github.com/openwrt/openwrt/pull/3037

lynxis commented on target/linux/ipq40xx/patches-5.4/706-dt-bindings-net-add-QCA807x-PHY.patch:
```diff
@@ -0,0 +1,61 @@
+From c66863c1ba8995b61e6d727d78a241c734f5bb57 Mon Sep 17 00:00:00 2001
```
Is there a reason why you did this as patch instead a file like you did for qca807x.c?

robimarko replied:
It's simply as it requires no changes that are OpenWrt specific, unlike the PHY driver.

lynxis commented on target/linux/ipq40xx/files/drivers/net/ethernet/qualcomm/essedma/edma.c:
```diff
@@ -2023,6 +2023,7 @@ int edma_open(struct net_device *netdev)
 	 */
 	if (adapter->poll_required) {
 		if (!IS_ERR(adapter->phydev)) {
+			genphy_soft_reset(adapter->phydev);
```
maybe add the commit description as comment here.

robimarko replied:
Good idea, will do.

robimarko replied:
Done

lynxis commented:
I'm unsure if we should squash all mdio commits into an atomic commit. So there are no "broken" commits in the tree.

Can you try to improve your commit messages.  E.g. https://who-t.blogspot.com/2009/12/on-commit-messages.html is a good example how to improve. I usually don't manage to shorten my commit summary file. But for the description it would be great to have ~80.
You have loooong lines in there and a lot of empty lines.

robimarko commented:
@f00b4r0 @lynxis Finally got around to cleaning this up quite a bit.
The indentation was all switched to 8 spaces per tab as checkpatch requires, ifdefs are not indented, SFP related conditional is removed as its unneeded and I would say 95% of checkpatch warnings and errors are resolved in the PHY and DT bindings patches.

I will clean up the remaining line or two that are over 100 characters per line today.

adschm commented on target/linux/ipq40xx/config-5.4:
```diff
@@ -414,6 +348,7 @@ CONFIG_PINCTRL_MSM=y
 # CONFIG_PINCTRL_SDM660 is not set
 # CONFIG_PINCTRL_SDM845 is not set
 # CONFIG_PINCTRL_SM8150 is not set
+CONFIG_PLUGIN_HOSTCC="g++"
```
Please manually drop the CONFIG_PLUGIN_HOSTCC again ...

robimarko replied:
Done

adschm commented on target/linux/ipq40xx/config-5.4:
```diff
@@ -299,7 +233,7 @@ CONFIG_MDIO_BITBANG=y
 CONFIG_MDIO_DEVICE=y
 CONFIG_MDIO_GPIO=y
-CONFIG_MDIO_IPQ40XX=y
+# CONFIG_MDIO_IPQ4019 is not set
```
And probably that change should not be in this patch?

robimarko replied:
Can I simply include it in the MDIO backport patch?
To me, it makes no sense as a standalone change

adschm replied:
Yes, that's what I meant.

robimarko replied:
Ok, I have already pushed the updated version with that.

robimarko committed:
ipq40xx: refresh kernel config

Generic kernel config changed a lot, so lets refresh ipq40xx
to reduce the diff.

Signed-off-by: Robert Marko <robert.marko@sartura.hr>

robimarko commented:
I have also updated the GPIO driver to not use magic values but rather mess with bits directly.
This is much cleaner and easier to see what is done.

robimarko commented:
I have resolved the final checkpatch warning and dropped the forgotten debug prints.
I hope that this is mergeable now.

f00b4r0 commented on target/linux/ipq40xx/files/drivers/net/phy/ar40xx.c:
```diff
@@ -2036,9 +1832,7 @@ static int ar40xx_probe(struct platform_device *pdev)
-	ret = register_switch(swdev, NULL);
-	if (ret)
-		goto err_unregister_phy;
+	register_switch(swdev, NULL);
```
Is it intentional to not check the return value here?

robimarko replied:
It's not, I forgot to add the check.
Its added now.

f00b4r0 commented on target/linux/ipq40xx/files/drivers/net/phy/qca807x.c:
```diff
@@ -0,0 +1,834 @@
+	case QCA807X_CDT_RESULTS_CROSS_SHORT_WITH_MDI3_SAME_SHORT:
+		return true;
+	}
+	return false;
```
(cosmetic) this should be the switch's `default` case

robimarko replied:
It's the same, all of the upstream drivers do it like this for PHY-s.
So I prefer to leave it this way.

f00b4r0 replied:
OK

f00b4r0 commented on target/linux/ipq40xx/files/drivers/net/phy/qca807x.c:
```diff
@@ -0,0 +1,834 @@
+		return err;
+
+	/* why bother the PHY if nothing can have changed */
+	if (phydev->autoneg == AUTONEG_ENABLE && old_link && phydev->link)
```
Disclaimer: I'm not familiar with phy drivers, but are you sure about this?
Wouldn't it be possible to have autoneg enabled, link stays present but remote party changed e.g. maximum advertised speed or duplex? If I'm not mistaken this can happen without requiring a link reset?
Also, don't you mean `&& old_link == phydev->link` here?

robimarko replied:
Well, this is just customized read_status from the at803x PHY driver as it has a similar PHY specific register telling the real link status.
The generic genphy_read_status() does similar things but on the generic Clause 22 register, so I presume that it's correct.
https://elixir.bootlin.com/linux/latest/source/drivers/net/phy/phy_device.c#L2342
Which does the same `phydev->autoneg == AUTONEG_ENABLE && old_link && phydev->link`

f00b4r0 replied:
Makes sense

f00b4r0 commented on target/linux/ipq40xx/files/drivers/net/phy/qca807x.c:
```diff
@@ -0,0 +1,834 @@
+			val = phy_read(phydev, QCA807X_MEDIA_SELECT_STATUS);
+			if (val & QCA807X_MEDIA_DETECTED_COPPER) {
+				qca807x_read_copper_status(phydev, true);
+			} else if ((val & QCA807X_MEDIA_DETECTED_1000_BASE_X) ||
+				   (val & QCA807X_MEDIA_DETECTED_100_BASE_FX)) {
```
This seems to exclude speed 10 which you still test for on line 545 above: is that intentional?

robimarko replied:
Yes, as this tests whether 1000Base-X or 100Base-FX fiber is detected, there is no 10Base fiber support.
But I left the test above as the PHY specific register has the same bits for speed status in the fiber page as in the copper one.

f00b4r0 replied:
OK, but it doesn't seem to make sense to have code testing for unsupported situations?

robimarko replied:
It has the bits for speed 10 documented in the datasheet, so why not have it?

f00b4r0 replied:
Depends whether there's hardware that actually support this. Otherwise you're just wasting space with code that will never be executed IMHO.

f00b4r0 replied:
Actually it would make more sense to have this if both copper/fiber were a single function, since they share quite a bit of code. But I can see reasons to keep them separated to.

adschm replied:
> Depends whether there's hardware that actually support this. Otherwise you're just wasting (a bit of) space with code that will never be executed IMHO.

Well, since we already have the "implementation" now, I would not actively make the driver "worse" by removing it again, unless we talk about huge amounts or substantially complex code.

f00b4r0 replied:
@adschm I'm not sure I follow? My point is there's a contradiction between two code paths, one which exclude the other from ever happening. Either there is hardware that supports 10, and the case must be treated in both paths, or there isn't, and it seems odd to keep dead code. My 2c though.

robimarko replied:
I mean, I can drop the case, it's only 2 lines though.
I intentionally made fiber and copper separate as fiber requires Clause 37 resolving of LPA while copper can use the generic one.
It also has a different PHY specific register page and this way it's much easier to follow the code and not get lost in a bunch of if-s.

robimarko replied:
Dropped the 10M check

f00b4r0 replied:
Sorry for being pedantic 😉

robimarko replied:
Nah, I am glad that somebody took a look at the code.
Saves me the pain with the mainteiners.

f00b4r0 commented on target/linux/ipq40xx/files/drivers/net/phy/qca807x.c:
```diff
@@ -0,0 +1,834 @@
+	int ss, err, page, lpa, old_link = phydev->link;
+
+	/* Only combo port has dual pages */
+	if (combo_port) {
```
From the only call site on line 579 below, this will always be true: maybe the code and function prototype can be simplified here?

robimarko replied:
Yeah, dropped the check as it only called on the combo port.

f00b4r0 reviewed (commented):
This looks much more readable, very nice job there!
Here are a few remarks, hopefully useful.

robimarko commented:
@adschm @f00b4r0 Any more comments?
Can this get merged as hAP ac2 depends on this and hAP ac2 is fully done.

f00b4r0 commented:
> @adschm @f00b4r0 Any more comments?
> 
> Can this get merged as hAP ac2 depends on this and hAP ac2 is fully done.

No further comments on my side, but I'm not a master of this subsystem or anything close to that :)

ynezz commented:
> PHY driver will be sent upstream soon.

So what's the state?

robimarko commented:
@ynezz I sent it upstream today.
Sorry for taking this long, but I had a lot of other stuff before vacation tomorrow.

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

robimarko commented:
Thanks,
but I dont see the commits in your staging tree.
